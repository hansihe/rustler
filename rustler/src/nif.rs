use crate::codegen_runtime::{c_char, c_int, c_uint, DEF_NIF_FUNC, NIF_ENV, NIF_TERM};

pub trait Nif {
    const NAME: *const c_char;
    const ARITY: c_uint;
    const FLAGS: c_uint;
    const FUNC: DEF_NIF_FUNC;
    const RAW_FUNC: unsafe extern "C" fn(
        nif_env: NIF_ENV,
        argc: c_int,
        argv: *const NIF_TERM,
    ) -> NIF_TERM;
}
