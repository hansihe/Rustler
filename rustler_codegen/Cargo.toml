[package]
name = "rustler_codegen"
description = "Compiler plugin for Rustler"
repository = "https://github.com/rusterlium/rustler/tree/master/rustler_codegen"
version = "0.21.0" # rustler_codegen version
authors = ["Hansihe <hansihe@hansihe.com>"]
license = "MIT/Apache-2.0"
edition = "2018"

[lib]
name = "rustler_codegen"
proc_macro = true

[dependencies]
syn = { version = "0.15.24", features = ["derive"] }
quote = "0.6.13"
heck = "0.3"
proc-macro2 = "0.4"
