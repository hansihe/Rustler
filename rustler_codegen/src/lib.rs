extern crate proc_macro;
use proc_macro::TokenStream;

extern crate syn;

#[macro_use]
extern crate quote;

mod util;
mod tuple;
mod map;

#[proc_macro_derive(NifMap)]
pub fn nif_map(input: TokenStream) -> TokenStream {
    let s = input.to_string();
    let ast = syn::parse_macro_input(&s).unwrap();
    let gen = map::transcoder_decorator(&ast);
    gen.unwrap().parse().unwrap()
}

#[proc_macro_derive(NifTuple)]
pub fn nif_tuple(input: TokenStream) -> TokenStream {
    let s = input.to_string();
    let ast = syn::parse_macro_input(&s).unwrap();
    let gen = tuple::transcoder_decorator(&ast);
    gen.unwrap().parse().unwrap()
}
