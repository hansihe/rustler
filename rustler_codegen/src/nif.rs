use proc_macro2::{Span, TokenStream};
use quote::{quote, ToTokens};
use syn::meta::ParseNestedMeta;
use syn::punctuated::Punctuated;
use syn::token::Comma;
use syn::LitStr;

const VALID_SCHEDULE_OPTIONS: [&str; 3] = ["Normal", "DirtyCpu", "DirtyIo"];

#[derive(Default)]
pub struct NifAttributes {
    schedule: Option<LitStr>,
    custom_name: Option<LitStr>,
}

impl NifAttributes {
    pub fn parse(&mut self, meta: ParseNestedMeta) -> syn::parse::Result<()> {
        if meta.path.is_ident("schedule") {
            let schedule: LitStr = meta.value()?.parse()?;

            if VALID_SCHEDULE_OPTIONS.contains(&schedule.value().as_str()) {
                self.schedule = Some(schedule);
                Ok(())
            } else {
                Err(meta.error(format!(
                    "The schedule option is expecting one of the values: {:?}",
                    VALID_SCHEDULE_OPTIONS
                )))
            }
        } else if meta.path.is_ident("name") {
            self.custom_name = Some(meta.value()?.parse()?);
            Ok(())
        } else {
            Err(meta.error("Unsupported nif macro attribute. Expecting schedule or name."))
        }
    }
}

pub fn transcoder_decorator(nif_attributes: NifAttributes, fun: syn::ItemFn) -> TokenStream {
    let sig = &fun.sig;
    let name = &sig.ident;
    let inputs = &sig.inputs;

    let flags = schedule_flag(nif_attributes.schedule);
    let function = fun.to_owned().into_token_stream();
    let arity = arity(inputs.clone());
    let decoded_terms = extract_inputs(inputs.clone());
    let argument_names = create_function_params(inputs.clone());
    let erl_func_name = nif_attributes
        .custom_name
        .map_or_else(|| name.to_string(), |n| n.value().to_string());

    if !erl_func_name.is_ascii() || erl_func_name.chars().any(|x| x.is_ascii_control()) {
        panic!("Only non-Control ASCII strings are supported as function names");
    }

    quote! {
        #[allow(non_camel_case_types)]
        pub struct #name;

        impl rustler::Nif for #name {
            const NAME: *const rustler::codegen_runtime::c_char = concat!(#erl_func_name, "\0").as_ptr() as *const rustler::codegen_runtime::c_char;
            const ARITY: rustler::codegen_runtime::c_uint = #arity;
            const FLAGS: rustler::codegen_runtime::c_uint = #flags as rustler::codegen_runtime::c_uint;
            const RAW_FUNC: unsafe extern "C" fn(
                nif_env: rustler::codegen_runtime::NIF_ENV,
                argc: rustler::codegen_runtime::c_int,
                argv: *const rustler::codegen_runtime::NIF_TERM
            ) -> rustler::codegen_runtime::NIF_TERM = {
                unsafe extern "C" fn nif_func(
                    nif_env: rustler::codegen_runtime::NIF_ENV,
                    argc: rustler::codegen_runtime::c_int,
                    argv: *const rustler::codegen_runtime::NIF_TERM
                ) -> rustler::codegen_runtime::NIF_TERM {
                    let lifetime = ();
                    let env = rustler::Env::new(&lifetime, nif_env);

                    let terms = std::slice::from_raw_parts(argv, argc as usize)
                        .iter()
                        .map(|term| rustler::Term::new(env, *term))
                        .collect::<Vec<rustler::Term>>();

                    fn wrapper<'a>(
                        env: rustler::Env<'a>,
                        args: &[rustler::Term<'a>]
                    ) -> rustler::codegen_runtime::NifReturned {
                        let result: std::thread::Result<_> = std::panic::catch_unwind(move || {
                            #decoded_terms
                            #function
                            Ok(#name(#argument_names))
                        });

                        rustler::codegen_runtime::handle_nif_result(result, env)
                    }
                    wrapper(env, &terms).apply(env)
                }
                nif_func
            };
            const FUNC: rustler::codegen_runtime::DEF_NIF_FUNC = rustler::codegen_runtime::DEF_NIF_FUNC {
                arity: Self::ARITY,
                flags: Self::FLAGS,
                function: Self::RAW_FUNC,
                name: Self::NAME
            };
        }
    }
}

fn schedule_flag(schedule: Option<LitStr>) -> TokenStream {
    let mut tokens = TokenStream::new();

    let flag = schedule.map_or("Normal".into(), |lit_str| lit_str.value());
    let flag_ident = syn::Ident::new(&flag, Span::call_site());

    tokens.extend(quote! { rustler::SchedulerFlags::#flag_ident });
    tokens
}

fn extract_inputs(inputs: Punctuated<syn::FnArg, Comma>) -> TokenStream {
    let mut tokens = TokenStream::new();
    let mut idx: usize = 0;

    for item in inputs.iter() {
        if let syn::FnArg::Typed(ref typed) = item {
            let name = &typed.pat;

            match &*typed.ty {
                syn::Type::Reference(typ) => {
                    let decoder = quote! {
                        let #name: #typ = match args[#idx].decode() {
                            Ok(value) => value,
                            Err(err) => return Err(err)
                        };
                    };

                    tokens.extend(decoder);
                }
                syn::Type::Path(syn::TypePath { path, .. }) => {
                    let typ = &typed.ty;
                    let ident = path.segments.last().unwrap().ident.to_string();

                    match ident.as_ref() {
                        "Env" => {
                            continue;
                        }
                        "Term" => {
                            let arg = quote! {
                                let #name: #typ = args[#idx];
                            };

                            tokens.extend(arg);
                        }
                        _ => {
                            let decoder = quote! {
                                let #name: #typ = match args[#idx].decode() {
                                    Ok(value) => value,
                                    Err(err) => return Err(err)
                                };
                            };

                            tokens.extend(decoder);
                        }
                    }
                }
                syn::Type::Tuple(typ) => {
                    let decoder = quote! {
                        let #name: #typ = match args[#idx].decode() {
                            Ok(value) => value,
                            Err(err) => return Err(err)
                        };
                    };

                    tokens.extend(decoder);
                }
                other => {
                    panic!("unsupported input given: {:?}", other);
                }
            }
        } else {
            panic!("unsupported input given: {:?}", stringify!(&item));
        };
        idx += 1;
    }

    tokens
}

fn create_function_params(inputs: Punctuated<syn::FnArg, Comma>) -> TokenStream {
    let mut tokens = TokenStream::new();

    for item in inputs.iter() {
        let name = if let syn::FnArg::Typed(ref typed) = item {
            &typed.pat
        } else {
            panic!("unsupported input given: {:?}", stringify!(&item));
        };

        tokens.extend(quote!(#name,));
    }

    tokens
}

fn arity(inputs: Punctuated<syn::FnArg, Comma>) -> u32 {
    let mut arity: u32 = 0;

    for (i, item) in inputs.iter().enumerate() {
        if let syn::FnArg::Typed(ref typed) = item {
            if let syn::Type::Path(syn::TypePath { path, .. }) = &*typed.ty {
                let ident = path.segments.last().unwrap().ident.to_string();

                if i == 0 && ident == "Env" {
                    continue;
                }

                if ident == "Env" {
                    panic!("Env must be the first argument in NIF functions");
                }
            };
        } else {
            panic!("unsupported input given: {:?}", stringify!(&item));
        };
        arity += 1;
    }

    arity
}
