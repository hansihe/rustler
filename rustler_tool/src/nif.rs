use libloading::{Library, Symbol};
use rustler_sys::ErlNifEntry;
use std::ffi::CStr;
use std::path::{Path, PathBuf};

pub struct Nif {
    pub name: String,
    pub arity: usize,
    pub flags: usize,
}

pub struct NifLibrary {
    pub path: PathBuf,
    pub name: String,
    pub nifs: Vec<Nif>,
}

#[cfg(unix)]
unsafe fn maybe_call_nif_init(
    lib: &Library,
) -> Result<*const ErlNifEntry, Box<dyn std::error::Error>> {
    let func: Symbol<unsafe extern "C" fn() -> *const ErlNifEntry> = lib.get(b"nif_init")?;

    Ok(func())
}

#[cfg(windows)]
unsafe fn maybe_call_nif_init(
    lib: &Library,
) -> Result<*const ErlNifEntry, Box<dyn std::error::Error>> {
    use rustler_sys::TWinDynNifCallbacks;
    static NULL_CALLBACKS: TWinDynNifCallbacks = TWinDynNifCallbacks {};
    let func: Symbol<unsafe extern "C" fn(*mut TWinDynNifCallbacks) -> *const ErlNifEntry> =
        lib.get(b"nif_init")?;

    func(&NULL_CALLBACKS)
}

impl NifLibrary {
    pub fn load(path: &Path) -> Result<NifLibrary, Box<dyn std::error::Error>> {
        unsafe {
            let lib = Library::new(path)?;
            let entry = maybe_call_nif_init(&lib)?;

            let name = CStr::from_ptr((*entry).name).to_str()?.to_string();
            let nif_array =
                std::slice::from_raw_parts((*entry).funcs, (*entry).num_of_funcs as usize);

            let mut nifs: Vec<_> = nif_array
                .iter()
                .filter_map(|f| {
                    Some(Nif {
                        name: CStr::from_ptr(f.name).to_str().ok()?.to_string(),
                        arity: f.arity as usize,
                        flags: f.flags as usize,
                    })
                })
                .collect();

            nifs.sort_by_key(|x| x.name.clone());

            Ok(NifLibrary {
                path: path.to_path_buf(),
                name,
                nifs,
            })
        }
    }
}
