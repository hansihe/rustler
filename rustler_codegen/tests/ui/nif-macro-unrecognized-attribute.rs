use rustler_codegen::nif;

#[nif(scheduler = "DirtyCpu")]
fn add(a: i64, b: i64) -> i64 {
    a + b
}

fn main() {}
