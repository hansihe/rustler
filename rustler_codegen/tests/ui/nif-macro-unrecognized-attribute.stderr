error: Unsupported nif macro attribute. Expecting schedule or name.
 --> tests/ui/nif-macro-unrecognized-attribute.rs:3:7
  |
3 | #[nif(scheduler = "DirtyCpu")]
  |       ^^^^^^^^^
